src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1632,13 +1483,8 @@ public Solution WithAdditionalDocumentTextLoader(DocumentId documentId, TextLoad
                 throw new ArgumentOutOfRangeException(nameof(mode));
             }
 
-            var newState = _state.UpdateAdditionalDocumentTextLoader(documentId, loader, mode);
-
-            // Note: state is currently not reused.
-            // If UpdateAdditionalDocumentTextLoader is changed to reuse the state replace this assert with Solution instance reusal.
-            Debug.Assert(newState != _state);
-
-            return new Solution(newState);
+            var newCompilationState = _compilationState.UpdateAdditionalDocumentTextLoader(documentId, loader, mode);
+            return newCompilationState == _compilationState ? this : new Solution(newCompilationState);
```

Comment:
this pattern was slightly different too.  With the presumption that updating text always creates something new.  That same logic is preserved, just pushed into the compilationstate method (including the assert).